src/main/java/io/cdap/plugin/aggregator/Pivot.java

```diff
@@ -0,0 +1,409 @@
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+
+/**
+ * Deduplicate aggregator.
+ */
+@Plugin(type = BatchAggregator.PLUGIN_TYPE)
+@Name("Pivot")
+@Description("Transpose specific rows as columns, and generate aggregate results in those columns")
+public class Pivot extends BatchReducibleAggregator<StructuredRecord, StructuredRecord, AggregateResult,
+  StructuredRecord> {
+  private final PivotConfig config;
+
+  @Nullable
+  private final Integer numPartitions;
+
+  private Schema outputSchema;
+  private List<PivotConfig.FunctionInfo> functionInfos;
+  private Map<String, Set<String>> pivotColumnsAndData;
+  //keep ordered columns name, to generate always same key from columns data models
+  private ArrayList<String> orderedColumnsName;
+  private List<String> columnsProduct;
+
+  private Map<String, String> fieldAliases;
+  private FailureCollector failureCollector;
+
+
+  public Pivot(PivotConfig config, @Nullable Integer numPartitions) {
+    this.config = config;
+    this.numPartitions = numPartitions;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    StageConfigurer stageConfigurer = pipelineConfigurer.getStageConfigurer();
+    failureCollector = stageConfigurer.getFailureCollector();
+
+    Schema inputSchema = stageConfigurer.getInputSchema();
+
+    config.validate(failureCollector);
+    if (inputSchema == null || config.containsMacro()) {
+      stageConfigurer.setOutputSchema(null);
+      return;
+    }
+
+    init(failureCollector);
+    validate(inputSchema, failureCollector);
+    outputSchema = generateOutputSchema(inputSchema, config.getPivotRow(), functionInfos, columnsProduct);
+    stageConfigurer.setOutputSchema(outputSchema);
+  }
+
+  @Override
+  public void prepareRun(BatchAggregatorContext context) throws Exception {
+    super.prepareRun(context);
+    if (numPartitions != null) {
+      context.setNumPartitions(numPartitions);
+    }
+    recordLineage(context);
+  }
+
+  @Override
+  public void initialize(BatchRuntimeContext context) {
+    init(context.getFailureCollector());
+    if (context.getInputSchema() != null) {
+      validate(context.getInputSchema(), context.getFailureCollector());
+      outputSchema = generateOutputSchema(context.getInputSchema(), config.getPivotRow(),
+                                          functionInfos, columnsProduct);
+    }
+  }
+
+  @Override
+  public void groupBy(StructuredRecord record, Emitter<StructuredRecord> emitter) {
+    String pivotRowFieldName = config.getPivotRow();
+
+    Schema.Field pivotRowField = record.getSchema().getField(pivotRowFieldName);
+    Object pivotRowValue = record.get(pivotRowFieldName);
+
+    Schema schema = Schema.recordOf(pivotRowField.getSchema().getRecordName() + ".unique",
```

Comment:
what is the record name if pivot row field is int/string

Reply:
it was null, fixed after refactor.